sw/SoC/examples/blinky/Makefile

```diff
@@ -1,88 +1,51 @@
-###############################################################################
-# Propject Info
+# Author: Stefano Mercogliano 	<stefano.mercogliano@unina.it>
+# Description:
+#   This Makefile defines the project name and paths for the common Makefile.
+#	Optionally, a user can define additional targets here.
 
-PROJECT = filename
-MCU = DEMOSYSTEM
+################
+# Program Name #
+################
 
-###############################################################################
+# $(MAKEFILE_LIST)			: Contains a list of all Makefiles parsed (This Makefile and common Makefile)
+# $(first $(MAKEFILE_LIST))	: Extracts the first entry, which is the current Makefile.
+# $(abspath ...)			: Converts the relative path of the Makefile to an absolute path.
+# $(dir ...)				: Extracts the directory portion of the absolute path.
+# $(shell basename ...)		: Extracts the last component of the directory path, which is the name of the directory containing the Makefile.
+PROGRAM_NAME = $(shell basename $(dir $(abspath $(firstword $(MAKEFILE_LIST)))))
```

Comment:
continuing from https://github.com/MaistoV/UninaSoC/pull/26#discussion_r1890057178

I think this is easy to break as well. Why don't we leverage our envvars, e.g. `SOC_SW_ROOT_DIR`?

Reply:
part of the path can be built using envvar ofc, but then we need the name of the specific project using that makefile. While we can combine the approaches, it would look clumsy imho. Moreover, we are not supposed to build a bare metal project using fancy flows. Instead, a user is likely to have a self-contained environment in its own project dir.

Reply:
While I agree that we must not get too clumsy, I think there are better approaches to do this. As calling `basename`, `dir`, `abspath` and `firstword` in a one-liner is a bit too much.

Maybe rethinking the include order, namley either:
1. include the common/Makefile in the beginning or middle of the project's one, not necessarily at the bottom
2. or, include the project's Makefile in the common's

Such alternatives would allow us to, e.g.:
1. Ditch the `PROGRAM_NAME` var for a wildcard expansion based on `.c` source name, rather than from the directory name
2. Define `PROGRAM_NAME` upfront, when calling the top `sw/SoC/Makefile ` and parsing path just like https://github.com/MaistoV/UninaSoC/blob/feature/linker_and_software/sw/SoC/Makefile#L21
3. Use alternative Makefile var definitions, e.g. recursive/deferred (=) vs simple/absolute (:=) variables

Reply:
These approaches do not seem to offer an improvement and would not be appropriately applicable (imho).

1.The primary idea is that a user should be able to define their own project while reusing most of our standard infrastructure. Consequently, including project-specific Makefiles into the common Makefile is not an option. Additionally, changing the inclusion order in the project's Makefile does not eliminate the need to explicitly define the project name for a specific project (unless the user manually specifies it).

2.Currently, there is no `.c` file named after the project directory. While replacing `main.c` with `dirname.c` is technically feasible, it would not be an elegant solution. Moreover, such an approach would still necessitate a "discovery" mechanism in the Makefile to identify the project's name.

3.The SoC Makefile is intended for quickly compiling examples. For custom projects, users are expected to rely on their own project-specific Makefiles.

It seems the difficulty in obtaining the project name arises from the assumption that the build process could be invoked arbitrarily from other directories. However, this is not the case; it is designed to be called exclusively from within the project's directory. Unless we opt to add a build_projects rule to the SoC Makefile or introduce other use cases that require invoking project Makefiles from outside the project directory, I think the current setup is sufficient. Alternatively, we could restore the previous approach that relied on the pwd command.